### ISSUE TITLE
Could this effort be consumed with any state-of-art GraphQL clients?

### ISSUE THREAD

**Issue by wey-gu** [Dragonchu/NebulaGraphQL#5]
Sorry I am not familiar with the GraphQL world, from my understanding, in Java, we could leverage this lib to enable GraphQL towards NebulaGraph, but does it mean I could universally leverage GraphQL clients from other lang/clients?

Or there are gap/opportunities here to create a GraphQL server based on/within this project?

Great work @Dragonchu !!!

Thanks!

**Comment by Dragonchu** [comment]
This is a library that converts graphql's json to nebula's ngql statement. GraphQL is based on the http protocol, we only need to get json from the http request to convert it to Ngql. For example, we can write a springboot service based on this library, and all client in other languages ​​only need to communicate based on the http protocol. We can directly use all existing GraphQL clients. Furthermore, GraphQL query mode and traditional REST query mode can exist at the same time.

**Comment by Dragonchu** [comment]
This project needs more contributors involved！

**Comment by wey-gu** [comment]
Got it @Dragonchu ,

Do you think a NebulaGraphQL-server implementation inside this project makes more sense or it's better to create another NebulaGraphQL-java-server project?

Also, is the existing mapping graph-pattern-complete? We could express most of the graph patterns now already or we need more effort here?

I think this is an exciting project that deserves more attention, I'd like to contribute and involve more mates here too, my limitation here is lack of java background though.

**Comment by Dragonchu** [comment]
> Do you think a NebulaGraphQL-server implementation inside this project makes more sense or it's better to create another NebulaGraphQL-java-server project?

In the initial design, I planned to create two projects. One is the current NebulaGraphQL project, which focuses on using [graphql-java](https://github.com/graphql-java/graphql-java) to implement the conversion of GraphQL syntax to ngql statements and returning query results according to requirements, among other core functionalities. The other project is NebulaGraphQL-spring, which may be based on the [spring-graphql](https://spring.io/projects/spring-graphql) project and aims to integrate Nebula's graphql queries into the spring ecosystem.NebulaGraphQL-spring will utilize NebulaGraphQL, but its primary focus is on providing a more convenient usage and better integration with the spring ecosystem, making it easier for developers to use.

**Comment by Dragonchu** [comment]
> Also, is the existing mapping graph-pattern-complete? We could express most of the graph patterns now already or we need more effort here?

So far, NebulaGraphQL has only implemented the simplest querying functionality and has done minimal prototyping to validate it. There is still a lot of work to be done, as a significant portion of query statement mapping remains incomplete. Additionally, we have concepts proposed by GraphQL such as [Authorization](https://graphql.org/learn/authorization/), [Pagination](https://graphql.org/learn/pagination/), and [Caching](https://graphql.org/learn/caching/) that need to be further enhanced and supplemented, providing a new development approach based on graph databases rather than traditional relational databases at the application layer.

**Comment by Dragonchu** [comment]
> I think this is an exciting project that deserves more attention, I'd like to contribute and involve more mates here too, my limitation here is lack of java background though.

@wey-gu Great!!! This project is currently relatively simple, without heavy dependencies on external components and frameworks.

**Comment by wey-gu** [comment]
> NebulaGraphQL-spring

Awesome!

From my understanding, this project be not only a server/tight integration for the spring ecosystem but also could be treated as a general graphql-server to enable plain REST/HTTP accessing from other clients?

**Comment by Dragonchu** [comment]
> > NebulaGraphQL-spring
> 
> Awesome!
> 
> From my understanding, this project be not only a server/tight integration for the spring ecosystem but also could be treated as a general graphql-server to enable plain REST/HTTP accessing from other clients?

Yep。This is what NebulaGraphQL-spring going to do.

**Comment by wey-gu** [comment]
> > Also, is the existing mapping graph-pattern-complete? We could express most of the graph patterns now already or we need more effort here?
> 
> So far, NebulaGraphQL has only implemented the simplest querying functionality and has done minimal prototyping to validate it. There is still a lot of work to be done, as a significant portion of query statement mapping remains incomplete. Additionally, we have concepts proposed by GraphQL such as [Authorization](https://graphql.org/learn/authorization/), [Pagination](https://graphql.org/learn/pagination/), and [Caching](https://graphql.org/learn/caching/) that need to be further enhanced and supplemented, providing a new development approach based on graph databases rather than traditional relational databases at the application layer.

Indeed, authN/authO/Pagination could be somehow wrapped in some level as there is such thing in NebulaGraph, while caching will be hard as in 3.x there is no cache(there will be one in 5.x!)
